services/venice-router/src/main/java/com/linkedin/venice/router/api/VeniceVersionFinder.java

```diff
@@ -98,119 +97,123 @@ public int getVersion(String storeName, BasicFullHttpRequest request) throws Ven
+    // new version not ready to serve
+    if (!newVersionPartitionResourcesReady) {
       stats.recordStalenessReason(StaleVersionReason.OFFLINE_PARTITIONS);
     }
-
-    if (!currentVersionDecompressorReady) {
-      errorMessage += " Decompressor not ready for current version (Has dictionary downloaded?).";
+    if (!newVersionDecompressorReady) {
       stats.recordStalenessReason(StaleVersionReason.DICTIONARY_NOT_DOWNLOADED);
     }
 
-    VersionStatus lastCurrentVersionStatus = store.getVersionStatus(lastCurrentVersion);
-    boolean prevVersionDecompressorReady = isDecompressorReady(store, lastCurrentVersion);
-    if (lastCurrentVersionStatus.equals(VersionStatus.ONLINE) && prevVersionDecompressorReady) {
-      String message = errorMessage + " Continuing to serve previous version: " + lastCurrentVersion + ".";
-      if (!EXCEPTION_FILTER.isRedundantException(message)) {
-        LOGGER.warn(message);
-      }
-      stats.recordStale(newCurrentVersion, lastCurrentVersion);
-      return lastCurrentVersion;
-    } else {
-      errorMessage += " Unable to serve previous version: " + lastCurrentVersion + ".";
-
-      if (!lastCurrentVersionStatus.equals(VersionStatus.ONLINE)) {
-        errorMessage += " Previous version has status: " + lastCurrentVersionStatus + ".";
-      }
-
-      if (!prevVersionDecompressorReady) {
-        errorMessage += " Decompressor not ready for previous version (Has dictionary downloaded?).";
-      }
-
-      /**
-       * When the router has only one available version, despite offline partitions, or dictionary not yet downloaded,
-       * it will return it as the available version.
-       * If the partitions are still unavailable or the dictionary is not downloaded by the time the records needs to
-       * be decompressed, then the router will return an error response.
-       */
-      String message = errorMessage + " Switching to serve new active version.";
-      if (!EXCEPTION_FILTER.isRedundantException(message)) {
-        LOGGER.warn(message);
-      }
-      lastCurrentVersionMap.put(storeName, newCurrentVersion);
-      stats.recordNotStale();
-      return newCurrentVersion;
+    /**
+     * When the router has only one available version, despite offline partitions, or dictionary not yet downloaded,
+     * it will return it as the available version.
+     * If the partitions are still unavailable or the dictionary is not downloaded by the time the records needs to
+     * be decompressed, then the router will return an error response.
+     */
+    // log if existing version ready to serve
+    Version existingVersion = store.getVersion(existingVersionNumber);
+    boolean existingVersionDecompressorReady =
+        isDecompressorReady(existingVersion, Version.composeKafkaTopic(storeName, existingVersionNumber));
+    // existing version ready to serve
+    if (!EXCEPTION_FILTER.isRedundantException(storeName)) {
```

Comment:
nit: make it redundant on isRedundantException(store_version) ? should be a minor as we do not expect store version changes are not that frequent.